tests/test_docopt.py

```diff
@@ -679,51 +714,271 @@ def test_issue_65_evaluate_argv_when_called_not_when_imported():
 
 
 def test_issue_71_double_dash_is_not_a_valid_option_argument():
-    with raises(DocoptExit):
+    with raises(DocoptExit, match=r"--log requires argument"):
         docopt("usage: prog [--log=LEVEL] [--] <args>...", "--log -- 1 2")
-    with raises(DocoptExit):
+    with raises(DocoptExit, match=r"-l requires argument"):
         docopt(
-            """usage: prog [-l LEVEL] [--] <args>...
-                  options: -l LEVEL""",
+            """\
+usage: prog [-l LEVEL] [--] <args>...
+options: -l LEVEL""",
             "-l -- 1 2",
         )
 
 
-usage = """usage: this
-
-usage:hai
-usage: this that
-
-usage: foo
-       bar
-
-PROGRAM USAGE:
- foo
- bar
-usage:
-\ttoo
-\ttar
-Usage: eggs spam
-BAZZ
-usage: pit stop"""
-
-
-def test_parse_section():
-    assert parse_section("usage:", "foo bar fizz buzz") == []
-    assert parse_section("usage:", "usage: prog") == ["usage: prog"]
-    assert parse_section("usage:", "usage: -x\n -y") == ["usage: -x\n -y"]
-    assert parse_section("usage:", usage) == [
-        "usage: this",
-        "usage:hai",
-        "usage: this that",
-        "usage: foo\n       bar",
-        "PROGRAM USAGE:\n foo\n bar",
-        "usage:\n\ttoo\n\ttar",
-        "Usage: eggs spam",
-        "usage: pit stop",
-    ]
-
-
-def test_issue_126_defaults_not_parsed_correctly_when_tabs():
-    section = "Options:\n\t--foo=<arg>  [default: bar]"
-    assert parse_defaults(section) == [Option(None, "--foo", 1, "bar")]
+option_examples: Sequence[tuple[str, Sequence[Option]]] = [
+    ("", []),
+    ("Some content\nbefore the first option.", []),
+    ("-f", [Option("-f", None, 0, False)]),
+    ("-f  Description.", [Option("-f", None, 0, False)]),
+    ("-f ARG  Description.", [Option("-f", None, 1, None)]),
+    ("-f ARG  Description. [default: 42]", [Option("-f", None, 1, "42")]),
+    ("--foo", [Option(None, "--foo", 0, False)]),
+    ("--foo  Description.", [Option(None, "--foo", 0, False)]),
+    ("--foo ARG  Description.", [Option(None, "--foo", 1, None)]),
+    ("--foo ARG  Description. [default: 42]", [Option(None, "--foo", 1, "42")]),
+    # Options can wrap over multiple lines
+    (
+        """\
+         \t --foo ARG, -f ARG  With a long
+
+         wrapped description
```

Comment:
@h4l what actually terminates an --option item? eg
```
--foo  Intended description.

Now I'm back to describing the usage of my program
in general. I include [default: 42] on accident
(OK this is very contrived)
```
should --foo have a default of 42? To me it looks like that option description ends on the single line and so it should not have that default.

Should we have the same behavior as when the usage: section terminates? eg the option description starts when we see `r"(?:  )|$"` (as we currently have), and then continues for any number of indented lines?
Oh but then this needs to be modified to "continues for any number of indented lines as long as they don't start with - or --"...yuck

This isn't really solving a likely problem, so we could just leave it unspecified.

Reply:
Yeah I think you answered it yourself, but the description continues until a line starts with `[ \t]-\S`. The previous and current code just splits the entire option description text block on lines starting in that way, and everything in between becomes part of the description for the option. So it's not at all nuanced in its handling of unusual stuff.

As for if the default should apply, I've certainly written option descriptions which extend over several lines. But I would keep them indented. Seems reasonable to end a description on a non-indented line, but it'd certainly require some thought.

Like I was mentioning in the other thread, the option description parsing code is really lenient. There was talk of defining a grammar for the docopt language in another issue — even if the implementation continued to use the hand-written parser, it'd be useful to have a grammar for the specification value of what is/isn't valid.